### ISSUE TITLE
auto-masking photo-metric loss

### ISSUE THREAD

**Issue by SmileyHu** [nianticlabs/monodepth2#103]
i would like to know where the masked photo-metric loss in the total losses, I didn't find the related code. does the "to_optimise" mean photo-metric loss rather than masked photo-metric loss?

**Comment by mrharicot** [comment]
You can find it here:

https://github.com/nianticlabs/monodepth2/blob/a053d0bc284a6945c268671d093c12ae857357e9/trainer.py#L447-L455

**Comment by SmileyHu** [comment]
i am confused,i think the predictive_mask is as Zhou et al,rather than the proposed auto-masking photometric loss in monodepth2.i would like to know where the auto-masking photometric loss, I didn't find the related code.

**Comment by mdfirman** [comment]
Hi @SmileyHu , 

The auto-masking happens in several places in the code, and I will go through where they are:

[Here](https://github.com/nianticlabs/monodepth2/blob/master/trainer.py#L436) is where the idendity reprojection losses are computed, i.e. the losses we would get by simply comparing the target image with the _unwarped_ source image.

These losses are concatenated together with the losses computed from the _warped_ source images [here](https://github.com/nianticlabs/monodepth2/blob/master/trainer.py#L471).

Finally, the minimum is taken over _all_ of these [here](https://github.com/nianticlabs/monodepth2/blob/master/trainer.py#L478).

Note that we are using a little trick here to speed things up: We could take a `min` once for automasking, and then a `min` again for minimum reprojection loss. Instead, though, we combine these two `min`s into one on [this line](https://github.com/nianticlabs/monodepth2/blob/master/trainer.py#L478).

Hope this helps!

**Comment by SmileyHu** [comment]
@mdfirman ,
Thank you very much. It does help me a lot, and i have one more question.
Here `loss += to_optimise.mean()`, i am confused  that why the "to_optimise"  did not multiply automasking?

**Comment by cswwp** [comment]
> Hi @SmileyHu ,
> 
> The auto-masking happens in several places in the code, and I will go through where they are:
> 
> [Here](https://github.com/nianticlabs/monodepth2/blob/master/trainer.py#L436) is where the identity reprojection losses are computed, i.e. the losses we would get by simply comparing the target image with the _unwarped_ source image.
> 
> These losses are concatenated together with the losses computed from the _warped_ source images [here](https://github.com/nianticlabs/monodepth2/blob/master/trainer.py#L471).
> 
> Finally, the minimum is taken over _all_ of these [here](https://github.com/nianticlabs/monodepth2/blob/master/trainer.py#L478).
> 
> Note that we are using a little trick here to speed things up: We could take a `min` once for automasking, and then a `min` again for minimum reprojection loss. Instead, though, we combine these two `min`s into one on [this line](https://github.com/nianticlabs/monodepth2/blob/master/trainer.py#L478).
> 
> Hope this helps!

@mdfirman Hi, i also confused by the identity_reprojection_losses, i print the ("color", frame_id, source_scale) is  ('color', 's', 0), where ('color', 's', 0) is located in inputs, so it means this identity_reprojection_losses represent the loss between origin left camera rgb image and origin right camera rgb image, Is my thinking right? If so, why you do this loss, as far as i know, this loss is difinitely exist in stereo camera and if we adjust params to reduce this loss, the disparity will be 0 according to stereo vision, we need disparity to beyond 0 so that predict depth.  I'm very confused by this issues, please reply me when you are convenient, thanks very much

`            if not self.opt.disable_automasking:
                identity_reprojection_losses = []
                for frame_id in self.opt.frame_ids[1:]:
                    print('("color", frame_id, source_scale):', ("color", frame_id, source_scale))
                    pred = inputs[("color", frame_id, source_scale)]
                    print('identity compute_reprojection_loss:', pred.shape,target.shape)
                    a = input()
                    identity_reprojection_losses.append(
                        self.compute_reprojection_loss(pred, target))

                identity_reprojection_losses = torch.cat(identity_reprojection_losses, 1)`

**Comment by xingyuuchen** [comment]
> @mdfirman , Thank you very much. It does help me a lot, and i have one more question. Here `loss += to_optimise.mean()`, i am confused that why the "to_optimise" did not multiply automasking?

I am also confused about  this. The auto-masking are computed here: `outputs["identity_selection/{}".format(scale)] = (idxs > identity_reprojection_loss.shape[1] - 1).float()`, but it is only used in visualization.

**Comment by SheffieldCao** [comment]
> @mdfirman , Thank you very much. It does help me a lot, and i have one more question. Here `loss += to_optimise.mean()`, i am confused that why the "to_optimise" did not multiply automasking?

My understanding is that, since the auto-masking mechanism wants to avoid the stationary pixels, the identity loss on these pixels is close to 0 due to the similar appearance. So the one step minimization can do all of the work. For example:
- on stationary pixels: `pe > identity_pe = 0`, so that `combine = identity_pe(after auto-masking)`;
- on dynamic pixels: `pe << identity_pe`, so that `combine = pe(after auto-masking)`;
